# Geocoding London neighbourhoods
This notebook is part of Coursera capstone project, but I separated it to keep readibility in the main project notebook. You can find the main project here: https://github.com/dedi400/Coursera_Capstone/blob/main/Capstone%20project%20-%20Location%20data.ipynb

The aim of this code is to get the list of London neighbourhoods from a wiki page and attach location data to each of them.

In [1]:
import pandas as pd
import googlemaps

## Reading data
Fortunately wikipedia puts all tables properly into a *\<table>* tag, so Pandas library can read it directly. *Read_html* returns a list of dataframes so we have to find the proper the table we need (it's the second table in this case).

In [2]:
london_url='https://en.wikipedia.org/wiki/List_of_areas_of_London'
london=pd.read_html(london_url,flavor='bs4')
len(london)

5

In [3]:
df_london=london[1]
print(df_london.shape)
df_london.head()

(533, 6)


Location                     London borough       Post town  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2    Addington                         Croydon[8]         CROYDON   
3   Addiscombe                         Croydon[8]         CROYDON   
4  Albany Park                             Bexley  BEXLEY, SIDCUP   

  Postcode district Dial code OS grid ref  
0               SE2       020    TQ465785  
1            W3, W4       020    TQ205805  
2               CR0       020    TQ375645  
3               CR0       020    TQ345665  
4         DA5, DA14       020    TQ478728

## Data cleaning
I have to rename the columns because *\&nbsp;* was used instead of normal space in th header and it makes trouble in filtering

In [4]:
df_london.columns=['Neighbourhood','Borough','Town','Postcode','Dial','OS']
df_london

Neighbourhood                            Borough            Town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
528         Woolwich                          Greenwich          LONDON   
529   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
530  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
531          Yeading                         Hillingdon           HAYES   
532         Yiewsley                         Hillingdon    WEST DRAYTON   

      Postcode Dial        OS  
0          SE2  020  TQ465785  
1       W3, W4  020  TQ205805  
2          CR0  020  TQ375645  
3          CR0  020  TQ345665  
4    DA5, DA14  020  TQ478728  
..         ...  ...       ...  
528       SE18  020  TQ435795  
529        KT4  020  TQ225655  
530        W12  020  TQ225815  
531        UB4  020  TQ115825  
532        UB7  020  TQ063804  

[533 rows x 6 columns]

In [5]:
df_london=df_london[df_london['Town']=='LONDON'].drop(['Town','Dial','OS'],axis=1)
df_london

Neighbourhood                            Borough  Postcode
0         Abbey Wood              Bexley, Greenwich [7]       SE2
1              Acton  Ealing, Hammersmith and Fulham[8]    W3, W4
6            Aldgate                           City[10]       EC3
7            Aldwych                    Westminster[10]       WC2
9            Anerley                        Bromley[11]      SE20
..               ...                                ...       ...
522       Wood Green                           Haringey       N22
523         Woodford                          Redbridge  IG8, E18
527    Woodside Park                             Barnet       N12
528         Woolwich                          Greenwich      SE18
530  Wormwood Scrubs             Hammersmith and Fulham       W12

[299 rows x 3 columns]

Removing references and whitespaces from Borough column

In [6]:
df_london['Borough']=df_london['Borough'].apply(lambda b: b.split('[',1)[0].strip())
df_london

Neighbourhood                         Borough  Postcode
0         Abbey Wood               Bexley, Greenwich       SE2
1              Acton  Ealing, Hammersmith and Fulham    W3, W4
6            Aldgate                            City       EC3
7            Aldwych                     Westminster       WC2
9            Anerley                         Bromley      SE20
..               ...                             ...       ...
522       Wood Green                        Haringey       N22
523         Woodford                       Redbridge  IG8, E18
527    Woodside Park                          Barnet       N12
528         Woolwich                       Greenwich      SE18
530  Wormwood Scrubs          Hammersmith and Fulham       W12

[299 rows x 3 columns]

I ran some more check on the neighbourhood list. It seems that postcodes don't follow neighbourhood borders so nice as in Toronto. As you can see below there are several different situations:

1. Some neighbourhoods have multiple postal codes (46 times)
2. Also, postal codes spans thorugh multiple neighbourhoods most of the time (220 out of 299!) Sometimes they span through multiple Boroughs (e.g. WC2 in Camden and Westminster).

In [7]:
print(df_london[df_london['Postcode'].str.contains(',')].shape)
df_london[df_london['Postcode'].str.contains(',')].sort_values('Borough')


(46, 3)


Neighbourhood                         Borough  \
108               Colney Hatch                          Barnet   
171                   Finchley                          Barnet   
60                 Brent Cross                          Barnet   
261               Kensal Green                           Brent   
320                    Neasden                           Brent   
127             Crystal Palace                         Bromley   
242                    Holborn                          Camden   
192                 Gospel Oak                          Camden   
375              Primrose Hill                          Camden   
18                    Barbican                            City   
457                     Temple               City, Westminster   
142                     Ealing                          Ealing   
1                        Acton  Ealing, Hammersmith and Fulham   
155                   Edmonton                         Enfield   
15                 Arnos Grove                         Enfield   
51   Blackheath Royal Standard                       Greenwich   
399                Shacklewell                         Hackney   
200                    Hackney                         Hackney   
438              Stamford Hill                         Hackney   
421            South Tottenham                        Haringey   
56                Bounds Green                        Haringey   
465                  Tottenham                        Haringey   
467             Tottenham Hale                        Haringey   
224                  Harringay                        Haringey   
10                       Angel                       Islington   
470               Tufnell Park               Islington, Camden   
143                Earls Court          Kensington and Chelsea   
337           North Kensington          Kensington and Chelsea   
243               Holland Park          Kensington and Chelsea   
416           South Kensington          Kensington and Chelsea   
471                 Tulse Hill                         Lambeth   
441                  Stockwell                         Lambeth   
355                       Oval                         Lambeth   
188                 Gipsy Hill                         Lambeth   
63                     Brixton                         Lambeth   
274                   Ladywell                        Lewisham   
455              Sydenham Hill             Lewisham, Southwark   
520                  Wimbledon                          Merton   
482                 Upton Park                          Newham   
504                   West Ham                          Newham   
523                   Woodford                       Redbridge   
328                  Newington                       Southwark   
157        Elephant and Castle                       Southwark   
284                     Leyton                  Waltham Forest   
425                Southfields                      Wandsworth   
289              Little Venice                     Westminster   

                Postcode  
108             N11, N10  
171          N2, N3, N12  
60              NW2, NW4  
261            NW10, NW6  
320            NW2, NW10  
127     SE19, SE20, SE26  
242             WC1, WC2  
192             NW5, NW3  
375        NW1, NW3, NW8  
18              EC1, EC2  
457             EC4, WC2  
142              W5, W13  
1                 W3, W4  
155              N9, N18  
15              N11, N14  
51             SE3, SE12  
399              E8, N16  
200  E5, E8, E9, N1, N16  
438             N15, N16  
421             N15, N17  
56              N11, N22  
465             N15, N17  
467             N15, N17  
224          N4, N8, N15  
10               EC1, N1  
470              N7, N19  
143            SW5, SW10  
337             W10, W11  
243         W8, W11, W14  
416        SW7, SW3, SW5  
471           SE24, SE27  
441             SW8, SW9  
355       SW8, SW9, SE1

In [8]:
df_london[df_london.duplicated(subset=['Postcode'],keep=False)].sort_values('Postcode')

Neighbourhood               Borough Postcode
492        Wapping         Tower Hamlets       E1
440        Stepney         Tower Hamlets       E1
381       Ratcliff         Tower Hamlets       E1
400       Shadwell         Tower Hamlets       E1
427   Spitalfields         Tower Hamlets       E1
..             ...                   ...      ...
269   King's Cross  Camden and Islington      WC1
431       St Giles                Camden      WC2
87   Charing Cross           Westminster      WC2
114  Covent Garden           Westminster      WC2
7          Aldwych           Westminster      WC2

[220 rows x 3 columns]

Based on these findings I decided that I am going to use Neighbourhood names instead of postal codes for this project.

There are 2-2 neighbourhoods with the same name, I'll add postcode to the name so geocoding can identify:

In [9]:
df_london[df_london.duplicated(subset=['Neighbourhood'],keep=False)].sort_values('Neighbourhood')

Neighbourhood   Borough Postcode
100    Church End     Brent     NW10
101    Church End    Barnet       N3
197    Grove Park  Hounslow       W4
198    Grove Park  Lewisham     SE12

In [10]:
to_be_adjusted=[100,101,197,198]
  
df_london.loc[to_be_adjusted,'Neighbourhood']=df_london.loc[to_be_adjusted,'Postcode']+' '+ \
                                                df_london.loc[to_be_adjusted,'Neighbourhood']

df_london.loc[to_be_adjusted]

Neighbourhood   Borough Postcode
100  NW10 Church End     Brent     NW10
101    N3 Church End    Barnet       N3
197    W4 Grove Park  Hounslow       W4
198  SE12 Grove Park  Lewisham     SE12

## Geocoding
I'm using Google Maps API for geocoding neighbourhoods, I found it more precise than OpenStreetMap when I search for areas instead of exact locations.

The googlemaps library returns a json-like data structure (actually it is a list of dictionaries of dictionaries). After some testing I found that `result[0]['geometry']['location']` gives a dictionary of `{'lat':'xxx','lng':'yyy'}` that is the coordinates we need.

In [11]:
google=googlemaps.Client(key='AIzaSyDJM_jXpm3PZcXGeLRZG9GiMeW1wz4DOzs') #to be deleted before publishing!

In [28]:
def neighbourhood_geocode(s):
    print(s['Neighbourhood'])
    result=google.geocode(s['Neighbourhood']+' London UK')
    try:
        return result[0]['geometry']['location']
    except:
        return

In [33]:
geo=df_london.apply(neighbourhood_geocode,axis=1,result_type='expand')

Abbey Wood
Acton
Aldgate
Aldwych
Anerley
Angel
Archway
Arnos Grove
Balham
Bankside
Barbican
Barnes
Barnsbury
Battersea
Bayswater
Bedford Park
Belgravia
Bellingham
Belsize Park
Bermondsey
Bethnal Green
Blackfriars
Blackheath
Blackheath Royal Standard
Blackwall
Bloomsbury
Bounds Green
Bow
Bowes Park
Brent Cross
Brent Park
Brixton
Brockley
Bromley (also Bromley-by-Bow)
Brompton
Brondesbury
Brunswick Park
Burroughs, The
Camberwell
Cambridge Heath
Camden Town
Canary Wharf
Cann Hall
Canning Town
Canonbury
Castelnau
Catford
Chalk Farm
Charing Cross
Charlton
Chelsea
Childs Hill
Chinatown
Chinbrook
Chingford
Chiswick
NW10 Church End
N3 Church End
Clapham
Clerkenwell
Colindale
Colliers Wood
Colney Hatch
Covent Garden
Cricklewood
Crofton Park
Crossness
Crouch End
Crystal Palace
Cubitt Town
Custom House
Dalston
Dartford
De Beauvoir Town
Denmark Hill
Deptford
Dollis Hill
Dulwich
Ealing
Earls Court
Earlsfield
East Dulwich
East Finchley
East Ham
East Sheen
Edmonton
Elephant and Castle
Eltham
Farringd

In [37]:
df_london=df_london.join(geo)
df_london.head()

Neighbourhood                         Borough Postcode        lat       lng
0    Abbey Wood               Bexley, Greenwich      SE2  51.492612  0.118818
1         Acton  Ealing, Hammersmith and Fulham   W3, W4  51.508372 -0.274440
6       Aldgate                            City      EC3  51.513438 -0.077171
7       Aldwych                     Westminster      WC2  51.513266 -0.117183
9       Anerley                         Bromley     SE20  51.411911 -0.067978

It seems all neighbourhoods are successfully geocoded:

In [44]:
df_london.isna().sum()

Neighbourhood    0
Borough          0
Postcode         0
lat              0
lng              0
dtype: int64

In [45]:
df_london[df_london['lat']==0]

Empty DataFrame
Columns: [Neighbourhood, Borough, Postcode, lat, lng]
Index: []

## Saving data for the main project
Before saving the dataframe to csv I rename columns to be consistent with other cities.

In [48]:
df_london.rename({'lat':'Lat','lng':'Lon'},axis=1,inplace=True)
df_london.head()

Neighbourhood                         Borough Postcode        Lat       Lon
0    Abbey Wood               Bexley, Greenwich      SE2  51.492612  0.118818
1         Acton  Ealing, Hammersmith and Fulham   W3, W4  51.508372 -0.274440
6       Aldgate                            City      EC3  51.513438 -0.077171
7       Aldwych                     Westminster      WC2  51.513266 -0.117183
9       Anerley                         Bromley     SE20  51.411911 -0.067978

In [49]:
df_london.to_csv('data/london_w_geocode.csv')